### DISCUSSION TITLE
Esp32 S3 is supported?

### DISCUSSION THREAD

**Discussion by mikadoryu** [corrados/edrumulus#72] [Q&A] [answered]
As the title says, it is possible to use an esp32-s3 (those with 2 usb c ports). I'm trying to compile the code but I have problems with the DAC since it can't find the library.

Searching I found that the S3 apparently do not have a normal dac like the previous esp32

**Comment by corrados** [comment]
Unfortunately, I do not have an ESP32-s3 available for testing.

> Searching I found that the S3 apparently do not have a normal dac like the previous esp32

What do you mean by that? Do you mean that the ADC in the ESP32-s3 must be controlled differently in software?

> but I have problems with the DAC since it can't find the library.

Sounds a bit like your Arduino IDE is not to be setup correctly. Have you installed the Espressif library like, e.g., [described here](https://randomnerdtutorials.com/installing-the-esp32-board-in-arduino-ide-windows-instructions/)?

**Reply by mikadoryu** [reply]
sorry for not being very clear.

Yes, I have the libraries installed correctly. Regarding the problem, it seems that the S3 do not have a built-in dac, so they do not need the DAC.h

I did not find much more about the problem apart from these two links that talk about the lack of the dac

https://github.com/espressif/esp-idf/issues/8119
https://www.esp32.com/viewtopic.php?t=22740

**Reply by corrados** [reply]
Is this the code in edrumulus_hardware.cpp which gives you compiler errors:
```
  // if the GIOP 25/26 are used, we have to set the DAC to 0 to get correct DC offset
  // estimates and reduce the number of large spikes
  dac_i2s_enable();
  dac_output_enable  ( DAC_CHANNEL_1 );
  dac_output_voltage ( DAC_CHANNEL_1, 0 );
  dac_output_disable ( DAC_CHANNEL_1 );
  dac_output_enable  ( DAC_CHANNEL_2 );
  dac_output_voltage ( DAC_CHANNEL_2, 0 );
  dac_output_disable ( DAC_CHANNEL_2 );
  dac_i2s_disable();
```
If yes, you could simply remove these lines and try again.

**Reply by mikadoryu** [reply]
First comment out the line in edrumulus_hardware.h` #include "driver/adc.h"` to avoid the library not found error.

After commenting the lines you say I get several errors:
![image](https://user-images.githubusercontent.com/8473292/228096388-336b78f5-d0af-44b2-b9fc-0bf97c96013c.png)
[error-log.txt](https://github.com/corrados/edrumulus/files/11084023/error-log.txt)

**Reply by corrados** [reply]
That is bad news. So, it seems the ESP32-S3 is no longer compatible to the hardware register definitions I use. Therefore, the ADC code has to be rewritten for the ESP32-S3.

**Reply by corrados** [reply]
FYI, I have just ordered an ESP32-S3 developer board in China. It may take up to two month until it arrives. As soon as I have it, I'll try to run Edrumulus on it.

**Reply by mikadoryu** [reply]
It's good that you want to review the errors with the S3.

One question I have is that the functions init_my_analogRead() and my_analogRead are because there was a delay problem in analogRead() do you know if this still happens in the libs for s3?

**Reply by corrados** [reply]
No, I don't. But I will find out :-)

**Reply by corrados** [reply]
I've now implemented a [fix for the ESP32-S3 compilation errors](https://github.com/corrados/edrumulus/commit/88bb64aaf6a72e10820937ee24cb11ba3821eac3). This I could do without having the hardware available. I now use the analogRead() function for the ESP32-S3 but I assume that it is way too slow for Edrumulus and that I have to implement my own analog read function similar to what exists right now for the old ESP32.
But at least you could give it a try now.

**Comment by corrados** [comment]
I just read in the Discussion overview that your question is still on "unanswered". So, I'll answer it now:

No, ESP32-S3 is not yet supported by Edrumulus. But as soon as I have one, I'll try to get Edrumulus running on that device :-).

**Comment by corrados** [comment, marked as answer]
An initial support for ESP32-S3 is done now. You can checkout the latest code on Git main branch and give it a try.
Up to now, I just checked that I get correct ADC readings from all configured inputs. I do not have the front end (i.e., the three resisters) available for the ESP32-S3, so I cannot do a test with a real pad. But by just touching the ADC connectors with my fingers gives me a lot of MIDI messages. So, in principle it should work now.

**Comment by corrados** [comment]
Today, I tested the ESP32-S3 with a real front end where my PD-85 mesh pad was connected and it worked just fine.

**Reply by thijstriemstra** [reply]
>  tested the ESP32-S3 .. it worked just fine.

Nice!

Looking forward to a USB host implementation where all midi processing is done on the S3, eg it shows up in our DAW as an Edrumulus USB MIDI device without any external software processing.

**Reply by corrados** [reply]
With the new edrumulus_gui.py which does not need Hairless MIDI and loopMIDI anymore, it is not really necessary to have USB host implementation. I think that the most important improvement of the S3 for Edrumulus is that we can have more inputs.

**Reply by peteallenm** [reply]
I'm really interested in building an edrumulus.. I built my own many years ago on a dsPIC (long since broken) but want to get the kit working again.
I'm looking at ordering a prototype 6 PCB, but figured I might as well do a couple of updates while I'm at it, and the S3 made sense. The 44pin dev boards have 18 exposed analogues, so I might add a few extra inputs. Would you recommend adding more dual zone inputs, more single zones, or a (non-jack based) breakout connector.
I'm kind of tempted to go to BAT56S's for the clamping diodes as it lowers the component count, but I don't think it would save space so it's probably not worth it

**Reply by corrados** [reply]
I cannot really recommend prototype 6. I built one but never really used it. Biggest drawback is that you cannot flash the board if it is connected to the prototype 6 PCB. You first have to disconnect it and flash it then.
And yes, a S3 makes more sense but I have not used an S3 in a real drum kit. I just verified with one input that it works at all. But it is definitely the more powerful device and if you want to use a lot of inputs, it makes totally sense.
Whether you want to use dual/single zone pads is totally up to you :-). Depends on how you want to setup your kit. E.g., a crash without a joke switch does not make any sense for me.
Regarding the diodes, that is a topic which was discussed a few times in this forum. I'm not a hardware design expert and cannot give you a good advice here. It will most probably work with using regular diodes (I did some tests with normal diodes and it just worked fine for me).

**Reply by peteallenm** [reply]
That's worth knowing about not being able to program. Is that because GPIO 0-3 are used for something?
I'll have a play with doing a new revision - I'd love to have the option of your advanced position sensing - that 3 piezo setup is something I've wanted to experiment with, and you've got it working well :)
A very rough layout shows an extra 2 dual inputs and 1 single input should all fit on a 200mm x 100mm board

**Reply by corrados** [reply]
> Is that because GPIO 0-3 are used for something?

Yes, that's the problem.

> I'd love to have the option of your advanced position sensing - that 3 piezo setup is something I've wanted to experiment with, and you've got it working well :)

Oh, that code is still very experimental. There are still a lot of TODOs in the code.

**Reply by 3hhh** [reply]
> I'm kind of tempted to go to BAT56S's for the clamping diodes as it lowers the component count, but I don't think it would save space so it's probably not worth it

The clamping diodes need to have a [forward voltage matching the operating range of the board](https://github.com/corrados/edrumulus/discussions/98#discussioncomment-7286101). You can of course go outside of the operating range and it _may_ work for a while, but your board may as well die once the clamping diodes kick in.

**Reply by peteallenm** [reply]
@corrados : Experimental code is all good - this already such a good platform to experiment with, so I'm very happy.
@3hhh : It's a good point. The bat54s (sorry, typo in first post) has a very very similar forward voltage, so it should be fine. It is fairly small SMD though so makes soldering a little tricky. But 2 diodes per package is nice and compact.

**Reply by peteallenm** [reply]
I've spun a board and sent it off to JLPCB (I went SMD in the end as it's easier to solder)
I've put it on my GitHub for now, and if it works I could do a pull request if it's helpful.

**Reply by corrados** [reply]
Sure :+1: Just name it "prototype_peteallenm" in the hardware directory.

**Comment by davy39** [comment]
Native USB is working with TinyUSB library. Se this PR #136

**Comment by mike76700** [comment]
when i compile with esp32 s3 i've got errors
with a normal esp32 everything ok

C:\Users\M_le_\Documents\Arduino\libraries\edrumulus-main\hardware.cpp: In member function 'void Edrumulus_hardware::init_my_analogRead()':
C:\Users\M_le_\Documents\Arduino\libraries\edrumulus-main\hardware.cpp:457:3: error: 'adc_power_acquire' was not declared in this scope
  457 |   adc_power_acquire();
      |   ^~~~~~~~~~~~~~~~~

**Reply by corrados** [reply]
Thanks for reporting. Should work now. Please try again.

**Comment by mike76700** [comment]
same errors with new file

**Reply by corrados** [reply]
That's not possible. I completely removed "adc_power_acquire()". Please check if you really use the latest Git code.

**Reply by mike76700** [reply]
sorry it's not the same
hardware.cpp: In member function 'void Edrumulus_hardware::init_my_analogRead()':
hardware.cpp:452:69: error: 'ADC_ATTEN_DB_12' was not declared in this scope
     adc1_config_channel_atten(static_cast<adc1_channel_t>(channel), ADC_ATTEN_DB_12);
                                                                     ^~~~~~~~~~~~~~~
hardware.cpp:452:69: note: suggested alternative: 'ADC_ATTEN_DB_11'
     adc1_config_channel_atten(static_cast<adc1_channel_t>(channel), ADC_ATTEN_DB_12);
                                                                     ^~~~~~~~~~~~~~~
                                                                     ADC_ATTEN_DB_11
*** [.pio\build\esp32-s3-devkitc-1\src\hardware.cpp.o] Error 1

**Reply by mike76700** [reply]
it seems to be ok with arduino ide ,
thanks a lot

**Reply by mike76700** [reply]
errors are with plateformio

**Reply by mike76700** [reply]
Hello, the flash with arduino ide worked well (not the case with platformio) full of noise even with resistors and capacitors and gpio ports that do not seem to be in the right place: nothing comes out of the 35 and 36 but on the side of gpio 4 and others sound comes out. a little complicated with esp32s3

**Reply by corrados** [reply]
Regarding PlatformIO, you might run into this problem: https://github.com/corrados/edrumulus/pull/113

**Reply by corrados** [reply]
Regarding noise: I would try to start with less inputs. Reduce the number of inputs to two or three in https://github.com/corrados/edrumulus/blob/c9d51d5923ce429b34ef1e81eae45334bd8e1140/hardware.cpp#L220.

Also, I would then try to increase the Threshold parameter until you do not get noise anymore.

**Reply by mike76700** [reply]
Thank you for your replies and for taking the time to answer me. I'm going to look at the leads you've given me.

**Reply by mike76700** [reply]
what should I change in the line of code for 1 pad only (snare drum)?

**Reply by corrados** [reply]
You can change it to, e.g.,

`*number_pins         = 1; // only one pad: snare`

**Reply by mike76700** [reply]
thank you, it seems better, noise only on the snare drum so attenuated with the treshold at 24 on octave(impossible to launch with python).the snare drum piezo very sensitive even at 2 in sensitivity(maybe a modification to do on my wiring)
thanks again, it's coming along

**Reply by mike76700** [reply]
sorry to ask so much of you 
to launch edrumulus_gui.py i get these errors (i thought of changing the com port for my esp32s3)
Traceback (most recent call last):
  File ‘C:\Users\M_le_\Downloads\edrumulus-main (1)\edrumulus-main\tools\edrumulus_gui.py’, line 696, in <module>
    ncurses_init()
  File ‘C:\Users\M_le_\Downloads\edrumulus-main (1)\edrumulus-main\tools\edrumulus_gui.py’, line 172, in ncurses_init
    mainwin = curses.initscr()
  File ‘C:\Users\M_le_AppData\Local\Programs\Python\Python312\Lib\curses\__init__.py’, line 30, in initscr
    fd=_sys.__stdout__.fileno())
AttributeError: ‘NoneType’ object has no attribute ‘fileno’
 I would like to know if it is possible for you to help me again. 
sorry again and thank you

**Reply by corrados** [reply]
Since it happens in the ncurses_init() call, this is obviously an issue with your Python ncurses installation. Maybe ncurses is not correctly installed with PIP or another method.

**Reply by mike76700** [reply]
reinstall windows-curses 2.3.3 now it's ok.
thank you for helping me

**Reply by mike76700** [reply]
for vitual port just see EdrumulusOut  but nothing for the in , is it because it's a esp32s3?

**Reply by mike76700** [reply]
and running the command with python edrumulus_gui.py rtmidi
I get these errors

Traceback (most recent call last):
  File “C:\Users\M_le_\edrumulus_gui.py”, line 661, in <module>
    midiin, port_name_in = open_midiinput([s for s in rtmidi.MidiIn().get_ports() if in_name in s][0], client_name=“EdrumulusGUI”)
                                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^
IndexError: list index out of range

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File “C:\Users\M_le_\edrumulus_gui.py”, line 665, in <module>
    raise Exception(“No native Edrumulus USB device (e.g., Teensy) nor loopMIDI driver found.”)
Exception: No native Edrumulus USB device (e.g., Teensy) nor loopMIDI driver found.

I have however configured loopmidi and hairless

**Reply by corrados** [reply]
You do not need loopmidi nor hairless midi anymore. You only need edrumulus_gui. I'll try it out on Windows now and check why it does not work with ESP32-S3.

**Reply by corrados** [reply]
I think you have to use the following command line in edrumulus/tools directory:

`python edrumulus_gui.py serial COM4`

The Edrumulus GUI started but it did not work correctly on my PC. Maybe it works on yours...

**Reply by corrados** [reply]
When I only enable one input (snare) it works for me. I simply started edrumulus GUI with "python edrumulus_gui.py serial COM4" and this connected correctly to the ESP32-S3 and created a MIDI port in Windows which you can use for your DAW:
![grafik](https://github.com/user-attachments/assets/4d96437f-3969-440e-9a6c-c865a4816015)

**Reply by corrados** [reply]
![grafik](https://github.com/user-attachments/assets/b4d45bcc-6651-4079-b435-e0801d668d6e)

**Reply by mike76700** [reply]
so I tried with com10 for me because my esp32s3 is on this port
it opens fine, it seems that a port is created but nothing comes out of my daw nor on midi-ox
with only 1 snare input

nothing comes out of this virtual port  
![capture ecran](https://github.com/user-attachments/assets/6c3b9652-79e4-4ffe-8eef-e28b53989419)

**Reply by corrados** [reply]
That is weird, why do you see MIDI note 40 in Edrumulus GUI whereas I see "38 snare". Maybe make a reset by pressing "r" key in Edrumulus GUI.

**Reply by mike76700** [reply]
it was a loose wire
i've got note 38 but still the same nothing comes out on the daw by virtual port

**Reply by corrados** [reply]
If you see notes in the Edrumulus GUI, we know that the communication between Edrumulus GUI and the ESP32-S3 works just fine. So, only the MIDI port creation under Windows seems to have an issue.

What I basically do is to create a MIDI out port in Windows with `pytemidi.Device("EdrumulusOut")` and send the MIDI data with `midiout.send(bytearray(serial_message))`. That works fine on my Windows.

Either your Windows system configuration does not allow to send MIDI or your pytemidi installation has a problem.

**Reply by mike76700** [reply]
reinstall pytemidi it's the same 
,anything to change ?
 perhaps uninstall loopmidi and hairless
i make a try in another computer as soon as possible to see if it's my computer.

**Reply by mike76700** [reply]
a way to use it without virtual port?
just for settings?

**Reply by corrados** [reply]
For settings, you can use Edrumulus GUI as you have it right now. As said above, the communication with the ESP32-S3 seems to work fine and therefore the settings adjustments should also work just fine.

**Comment by mike76700** [comment]
can i use it without virtual port creation?

**Reply by corrados** [reply]
Why do you want to remove it? Do you get issues in your system if it appears? I thought that you could simply ignore it if you do not need to use it.
